src/plugins/vis_type_vega/public/data_model/vega_parser.ts

```diff
@@ -157,25 +155,97 @@ The URL is an identifier only. Kibana and your browser will never access this UR
     this._parseControlPlacement();
     if (this.useMap) {
       this.mapConfig = this._parseMapConfig();
-    } else if (this.spec && this.spec.autosize === undefined) {
-      // Default autosize should be fit, unless it's a map (leaflet-vega handles that)
-      this.spec.autosize = { type: 'fit', contains: 'padding' };
+      this.useResize = false;
+    } else if (this.spec) {
+      this._compileWithAutosize();
     }
 
     await this._resolveDataUrls();
 
     if (this.isVegaLite) {
       this._compileVegaLite();
     }
+  }
+
+  /**
+   * Ensure that Vega and Vega-Lite will take the full width of the container unless
+   * the user has explicitly disabled this setting. Also sets the default width to include
+   * the padding. This creates the least configuration needed for most cases, with the option
+   * to do more.
+   */
+  private _compileWithAutosize() {
+    const defaultAutosize = {
+      type: 'fit',
+      contains: 'padding',
+    };
+
+    if (
+      !this.isVegaLite &&
+      this.spec.autosize &&
+      typeof this.spec.autosize === 'object' &&
+      'signal' in this.spec.autosize
+    ) {
+      // Vega supports dynamic autosize information, so we ignore it
+      return;
+    }
+
+    if (!this.spec.autosize && typeof this.spec.autosize !== 'undefined') {
+      this._onWarning(
+        i18n.translate('visTypeVega.vegaParser.autoSizeDoesNotAllowFalse', {
+          defaultMessage:
+            '{autoSizeParam} is enabled, it can only be disabled by setting {autoSizeParam} to {noneParam}',
+          values: {
+            autoSizeParam: '"autosize"',
+            noneParam: '"none"',
+          },
+        })
+      );
+
+      this.spec.autosize = defaultAutosize;
+      this.useResize = true;
+    } else if (typeof this.spec.autosize === 'string') {
```

Comment:
nit: maybe we can use `switch` intead of `if`? I think code will be a little bit easier for understanding
```
  private _compileWithAutosize() {
    const defaultAutosize = {
      type: 'fit',
      contains: 'padding',
    };

    let autosize: any = this.spec.autosize;
    let useResize = true;

    if (!this.isVegaLite && autosize && typeof autosize === 'object' && 'signal' in autosize) {
      // Vega supports dynamic autosize information, so we ignore it
      return;
    }

    switch (typeof autosize) {
      case 'string':
        useResize = autosize !== 'none';
        autosize = { ...defaultAutosize, type: autosize };
        break;
      case 'object':
        autosize = { ...defaultAutosize, ...autosize } as {
          type: string;
          contains: string;
        };
        useResize = Boolean(autosize?.type && autosize?.type !== 'none');
        break;
      default:
        if (!autosize && typeof autosize !== 'undefined') {
          this._onWarning(
            i18n.translate('visTypeVega.vegaParser.autoSizeDoesNotAllowFalse', {
              defaultMessage:
                '{autoSizeParam} is enabled, it can only be disabled by setting {autoSizeParam} to {noneParam}',
              values: {
                autoSizeParam: '"autosize"',
                noneParam: '"none"',
              },
            })
          );
        }
        autosize = defaultAutosize;
    }

    if (
      useResize &&
      ((this.spec.width && this.spec.width !== 'container') ||
        (this.spec.height && this.spec.height !== 'container'))
    ) {
      this._onWarning(
        i18n.translate('visTypeVega.vegaParser.widthAndHeightParamsAreIgnored', {
          defaultMessage:
            '{widthParam} and {heightParam} params are ignored because {autoSizeParam} is enabled. Set {autoSizeParam}: {noneParam} to disable',
          values: {
            widthParam: '"width"',
            heightParam: '"height"',
            autoSizeParam: '"autosize"',
            noneParam: '"none"',
          },
        })
      );
    }
    if (useResize) {
      this.spec.width = 'container';
      this.spec.height = 'container';
    }

    this.spec.autosize = autosize;
    this.useResize = useResize;
  }
```

Reply:
Thanks for the suggestion! I've taken some of the refactoring you suggested without actually implementing a `switch` statement.